### PR TITLE
adding openmmtorchplugin 

### PR DESCRIPTION
**Pull request by SCMusson**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by SCMusson** [comment]
@conda-forge-admin please ping conda-forge/help-c-cpp 

Any help here would be appreciated.

File: recipes/openmmtorchplugin/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+      thisexitcode=$?
+      summary+="\n${f}: "
+      if [[ $thisexitcode == 0 ]]; then summary+="OK"; else summary+="FAILED"; fi
+      ((exitcode+=$thisexitcode))
+    done
+    echo "-------"
+    echo "Summary"
+    echo "-------"
+    echo -e "${summary}"
+    exit $exitcode
```
Comment (carterbox) [inline recipes/openmmtorchplugin/meta.yaml:76]: Please move this to `run_test.sh`

https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#run-test-script

File: recipes/openmmtorchplugin/conda_build_config.yaml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+pytorch_version:
+  - 1.11
```
Comment (carterbox) [inline recipes/openmmtorchplugin/conda_build_config.yaml:2]: pytorch is already pinned to 1.11 for all packages on conda-forge. Use that variable instead.

https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml

File: recipes/openmmtorchplugin/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+build:
+  number: 0
+  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}   # [cuda_compiler_version != "None"]
+  string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                      # [cuda_compiler_version == "None"]
+  skip: true  # [not linux or (linux and cuda_compiler_version in (undefined, 'None', '10.2'))]
+  rpaths:
+    - lib/
+    - {{ SP_DIR }}/torch/lib
+  missing_dso_whitelist:
+    - '*/libcuda.*'  # [linux64 or ppc64le]
```
Comment (carterbox) [inline recipes/openmmtorchplugin/meta.yaml:21]:
```suggestion
```
You should add the cudatoolkit to the host section of your requirements, so this whitelist is not necessary. Use a selector so it is not required for the noncuda build.

File: recipes/openmmtorchplugin/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+source:
+  url: https://github.com/SCMusson/{{ name }}/archive/v{{ version }}.tar.gz
+  sha256: 1a9197c32209eaa37b9dcb1a7ef6ed30841ea5beed124b902485c20f5dd92021
+
+build:
+  number: 0
+  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}   # [cuda_compiler_version != "None"]
+  string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                      # [cuda_compiler_version == "None"]
+  skip: true  # [not linux or (linux and cuda_compiler_version in (undefined, 'None', '10.2'))]
```
Comment (carterbox) [inline recipes/openmmtorchplugin/meta.yaml:16]:
```suggestion
  skip: true  # [cuda_compiler_version in (undefined, 'None', '10.2')]
  skip: true  # [not linux]
```
I believe that multiple skips is equivalent to logical OR.

File: recipes/openmmtorchplugin/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    - openmm
+    - pytorch-gpu {{ pytorch_version }}  # [cuda_compiler_version not in (undefined, 'None')]
+    - pytorch-cpu {{ pytorch_version }}  # [osx]
+    - ocl-icd  # [linux]
+    - khronos-opencl-icd-loader  # [osx]
+  run:
+    - python
+    - pytorch
+    - {{ pin_compatible('pytorch-gpu', max_pin='x.x.x') }}  # [cuda_compiler_version not in (undefined, 'None')]
+    - {{ pin_compatible('pytorch-cpu', max_pin='x.x.x') }}  # [osx]
```
Comment (carterbox) [inline recipes/openmmtorchplugin/meta.yaml:45]:
```suggestion
    - {{ pin_compatible('pytorch-gpu', max_pin='x.x.x') }}  # [cuda_compiler_version not in (undefined, 'None')]
    - {{ pin_compatible('pytorch-cpu', max_pin='x.x.x') }}  # [osx]
```
Redundant package spec.

**Review by carterbox — CHANGES_REQUESTED** [review]
1. You are shipping C headers. Is this package intended to provide a C interface for other packages to link against or is it only supposed to be a python module?

2. You are shipping test files in /share. Why? Do end users need these to use your plugin?

3. OpenMM expects your plugins to be located in /lib/plugins?

File: recipes/openmmtorchplugin/build.sh [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+fi
+
+# Build in subdirectory and install.
+mkdir -p build
+cmake --version
+cd build
+cmake ${CMAKE_ARGS} ${CMAKE_FLAGS} ${SRC_DIR}
+make -j$CPU_COUNT install
+
+cp ${SRC_DIR}/python/myOpenMMSwigHeaders.i ${PREFIX}/include/swig/
```
[the remark this answers is not in the document]
Comment (SCMusson) [reply recipes/openmmtorchplugin/build.sh:31]: > 1. You are shipping C headers. Is this package intended to provide a C interface for other packages to link against or is it only supposed to be a python module?

I'm intending to provide a C interface. This seems to be standard practice for OpenMM plugins to have a high level python interface and a low level C interface. 
> 
> ~2. You are shipping test files in /share. Why? Do end users need these to use your plugin?~ See comment below about copying test files separately.

This seems to be standard practice for OpenMM plugins. I used this feedstock as a template: https://github.com/conda-forge/openmm-torch-feedstock/blob/main/recipe/meta.yaml 

>     3. OpenMM expects your plugins to be located in /lib/plugins?

I'm building everything as the OpenMM team have built their own plugins. Including using their conda-forge pull request/feedstock for their openmm-torch plugin as template for this one.

File: recipes/openmmtorchplugin/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    - ocl-icd  # [linux]
+    - ocl-icd-system  # [linux]
+    - khronos-opencl-icd-loader  # [osx]
+    - ocl_icd_wrapper_apple      # [osx]
+
+test:
+  imports:
+    - torchintegratorplugin
+  commands: |
+    cd ${CONDA_PREFIX}/share/{{ name }}/tests
```
[the remark this answers is not in the document]
Comment (SCMusson) [reply recipes/openmmtorchplugin/meta.yaml]: This seems to be standard practice for OpenMM plugins. I used this feedstock as a template: https://github.com/conda-forge/openmm-torch-feedstock/blob/main/recipe/meta.yaml

**Comment by SCMusson** [comment]
> 1. You are shipping C headers. Is this package intended to provide a C interface for other packages to link against or is it only supposed to be a python module?

I'm intending to provide a C interface. This seems to be standard practice for OpenMM plugins to have a high level python interface and a low level C interface. 
> 
> ~2. You are shipping test files in /share. Why? Do end users need these to use your plugin?~ See comment below about copying test files separately.

This seems to be standard practice for OpenMM plugins. I used this feedstock as a template: https://github.com/conda-forge/openmm-torch-feedstock/blob/main/recipe/meta.yaml 

>     3. OpenMM expects your plugins to be located in /lib/plugins?

I'm building everything as the OpenMM team have built their own plugins. Including using their conda-forge pull request/feedstock for their openmm-torch plugin as template for this one.

File: recipes/openmmtorchplugin/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+  sha256: 1a9197c32209eaa37b9dcb1a7ef6ed30841ea5beed124b902485c20f5dd92021
+
+build:
+  number: 0
+  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}   # [cuda_compiler_version != "None"]
+  string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                      # [cuda_compiler_version == "None"]
+  skip: true  # [not linux or (linux and cuda_compiler_version in (undefined, 'None', '10.2'))]
+  rpaths:
+    - lib/
+    - {{ SP_DIR }}/torch/lib
```
[the remark this answers is not in the document]
Comment (SCMusson) [reply recipes/openmmtorchplugin/meta.yaml:20]: Everything breaks when I remove it. It is present in the openmm-torch https://github.com/conda-forge/openmm-torch-feedstock/blob/main/recipe/meta.yaml plugin  built and maintained by the openmm team so I copied it across and it fixed the issues I was having.

File: recipes/openmmtorchplugin/conda_build_config.yaml (hunk 2 above)
Comment (SCMusson) [reply recipes/openmmtorchplugin/conda_build_config.yaml]: I have removed this as requested

File: recipes/openmmtorchplugin/meta.yaml (hunk 1 above)
Comment (SCMusson) [reply recipes/openmmtorchplugin/meta.yaml]: I this removed this as requested

**Comment by SCMusson** [comment]
The one failing check looks like a memory issue with the azure server rather than with package.

File: recipes/openmmtorchplugin/meta.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (carterbox) [reply recipes/openmmtorchplugin/meta.yaml:20]: That recipe does not use the `build/rpaths` section? I think what you are saying is that you copied files from `openmm-torch` into your package. Instead, you should add `openmm-torch` to the `requirements/host` and/or `requirements/run` sections of your recipe.

[the remark this answers is not in the document]
Comment (SCMusson) [reply recipes/openmmtorchplugin/meta.yaml:20]: My point still stands, If I remove that it breaks and I don't know how to fix it.

[the remark this answers is not in the document]
Comment (carterbox) [reply recipes/openmmtorchplugin/meta.yaml:20]: You binaries are linking to `libtorch.so`, but this library is located in a non-standard place, so it needs to be added to the path manually.

File: recipes/openmmtorchplugin/meta.yaml (hunk 3 above)
Comment (carterbox) [reply recipes/openmmtorchplugin/meta.yaml]: You don't actually link against any of the CUDA toolkit libraries, but this whitelist entry is not needed.

**Comment by carterbox** [comment]
> I'm intending to provide a C interface. This seems to be standard practice for OpenMM plugins to have a high level python interface and a low level C interface.

Typically for packages with a C interface, we would split the python/c parts into separate outputs because the APIs could be used separately. However, your C interface is a pytorch module right? So anyone who builds against your C API is also building a pytorch module?

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by jaimergp** [comment]
Isn't this available already at https://github.com/conda-forge/openmm-torch-feedstock?

File: recipes/openmmtorchplugin/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+source:
+  url: https://github.com/SCMusson/{{ name }}/archive/v{{ version }}.tar.gz
+  sha256: 1a9197c32209eaa37b9dcb1a7ef6ed30841ea5beed124b902485c20f5dd92021
+
+build:
+  number: 0
+  string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}   # [cuda_compiler_version != "None"]
+  string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                                 # [cuda_compiler_version == "None"]
+  skip: true  # [cuda_compiler_version in (undefined, 'None', '10.2')]
+  skip: true  # [not linux]
```
[the remark this answers is not in the document]
Comment (carterbox) [reply recipes/openmmtorchplugin/meta.yaml]: No need to remove the selectors, I just wanted to check that skipping osx wasn't a mistake.

**Comment by carterbox** [comment]
Typically for packages with a C interface, we would split the python/c parts into separate outputs because the APIs could be used separately. However, your C interface is a pytorch module right? So anyone who builds against your C API is also building a pytorch module?

**Comment by SCMusson** [comment]
Apologies for my sluggish replies.


> Typically for packages with a C interface, we would split the python/c parts into separate outputs because the APIs could be used separately. However, your C interface is a pytorch module right? So anyone who builds against your C API is also building a pytorch module?

This OpenMM plugin interfaces pytorch and OpenMM, I can't imagine anyone building against the C api and not intending to use pytorch/openmm.


Do you have any remaining specific changes in mind you wish me to address? I currently slightly unsure of how you wish me to proceed from here.

File: recipes/openmmtorchplugin/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}  # [cuda_compiler_version not in (undefined, 'None')]
+    - {{ cdt('mesa-libgl-devel') }}  # [linux]
+    - cmake
+    - make
+    - swig
+  host:
+    - khronos-opencl-icd-loader  # [osx]
+    - ocl-icd  # [linux]
```
Comment (carterbox) [inline recipes/openmmtorchplugin/meta.yaml:39]: The [link inspector reports ](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=517897&view=logs&j=1c012e3e-158a-5e5c-4fba-bceacec9f3ac&t=a2ee1cdc-f13c-52b8-e778-ce5a3d165825&l=2222)that ocl-icd is unused by your package. Is this copypasta?

File: recipes/openmmtorchplugin/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    - ocl-icd-system  # [linux]
+    - ocl_icd_wrapper_apple  # [osx]
+    - python
+  run_constrained:
+    - pytorch =*={{ torch_proc_type }}*
+
+test:
+  imports:
+    - torchintegratorplugin
+
```
Comment (carterbox) [inline recipes/openmmtorchplugin/meta.yaml:59]:
```suggestion
test:
  imports:
    - torchintegratorplugin
  commands:
    - test -f $PREFIX/include/TorchIntegratorKernels.h  # [unix]
```
Test should cover all types of files shipped.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/openmmtorchplugin/meta.yaml (hunk 10 above)
Comment (SCMusson) [reply recipes/openmmtorchplugin/meta.yaml]: I removed it and it still works so yes

**Comment by SCMusson** [comment]
@carterbox anything further you wish to change?

**Comment by SCMusson** [comment]
@carterbox Any further changes you wish to propose?

File: recipes/openmmtorchplugin/meta.yaml (hunk 10 above)
Comment (carterbox) [reply recipes/openmmtorchplugin/meta.yaml]: You forgot the ones in the run section. If you aren't linking against ocl, then your package probably doesn't need any ocl components are runtime either.

Comment (SCMusson) [reply recipes/openmmtorchplugin/meta.yaml]: Ah sorry, I've removed the rest

**Pull request merged**